doc/languages-frameworks/python.section.md

```diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
+of development: individual projects create their own python environments without
+impacting the global environment or each other.
 
-  packageOverrides = pkgs: with pkgs; {
-    myEnv = python38.withPackages (ps: with ps; [ numpy toolz ]);
-  };
-}
+#### Ad-hoc temporary python environment with `nix-shell`
+
+The simplest way to start playing with the way nix wraps and sets up python
+environments is with `nix-shell` at the cmdline. These environments create a
+temporary shell session with a python and a *precise* list of packages (plus
+their runtime dependencies), with no other python packages in the python
+interpreter's scope.
+
+To create a Python 3.8 session with `numpy` and `toolz` available, run:
+
+```sh
+$ nix-shell -p 'python38.withPackages(ps: with ps; [ numpy toolz ])'
 ```
-and install it in your profile with
 
-```shell
-nix-env -iA nixpkgs.myEnv
+By default `nix-shell` will start a `bash` session with this interpreter in our
+`PATH`, so if we then run:
+
+```
+[nix-shell:~/src/nixpkgs]$ python3
+Python 3.8.1 (default, Dec 18 2019, 19:06:26)
+[GCC 9.2.0] on linux
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import numpy; import toolz
 ```
 
-The environment is is installed by referring to the attribute, and considering
-the `nixpkgs` channel was used.
+Note that no other modules are in scope, even if they were imperatively
+installed into our user environment as a dependency of a python application:
 
-##### Environment defined in `/etc/nixos/configuration.nix`
+```
+>>> import requests
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+ModuleNotFoundError: No module named 'requests'
+```
 
-For the sake of completeness, here's another example how to install the
-environment system-wide.
+We can add as many additional modules onto the `nix-shell` as we need, and we
+will still get 1 wrapped python interpreter. We can start the interpreter
+directly like so:
 
-```nix
-{ # ...
+```sh
+$ nix-shell -p 'python38.withPackages(ps: with ps; [ numpy toolz requests ])' --run python3
+these derivations will be built:
+  /nix/store/xbdsrqrsfa1yva5s7pzsra8k08gxlbz1-python3-3.8.1-env.drv
+building '/nix/store/xbdsrqrsfa1yva5s7pzsra8k08gxlbz1-python3-3.8.1-env.drv'...
+created 277 symlinks in user environment
+Python 3.8.1 (default, Dec 18 2019, 19:06:26)
+[GCC 9.2.0] on linux
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import requests
+>>>
+```
 
-  environment.systemPackages = with pkgs; [
-    (python38.withPackages(ps: with ps; [ numpy toolz ]))
-  ];
-}
+Notice that this time it built a new python environment, which now includes
+`requests`. This is just creating wrapper scripts that setup `NIX_PYTHONPATH`,
+but re-using the actual modules themselves, so it's fast and efficient -- even
+if we're adding "big" modules like `pytorch` or `tensorflow`.
+
+To find the names of python attributes, startup `nix repl .` in a Nixpkgs
```

Comment:
I guess `nix repl '<nixpkgs>'` would be simpler, since the user doesn't need a `nixpkgs` checkout and they'll see packages from their currently configured channel.

Reply:
They need a channel, though.  Personally I've always found channels to be a confusing stumbling block for beginners; most developers can intuitively grasp a git repo clone, and git pull / checkout to change their "channel", and can reason about it: e.g., I'm getting "the exact python expressions from *this directory's* python versions".  

At any rate, it's irrelevant in this case, since it seems we're purging reference to nix 2.0 cmds in the manual for the time being.